javascript/photo.js

```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
+     * display
      */
     display: function(numberOfImages, cacheSize) {
-        var count = localStorage.getItem("count");
-        if (count === null) {
+        if (!this.isValidLocalStorageState(cacheSize)) {
             this.setup();
         }
-        count = Number(count);
-        var cacheIndex = Number(localStorage.getItem("cacheIndex"));
+        // From this point onwards, count, order and cachedImages are in a valid state (to know what a valid state is, refer to the IsValidLocalStorage function).
+        var count = Number(localStorage.getItem("count"));
         var order = JSON.parse(localStorage.getItem("order"));
-        if (count < this.STARTUP_IMAGE_COUNT) {
-            this.displayImage(this.offlineImages[order[count]]);
+        var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
+        // We display backup images in the following cases.
+        // * count < STARTUP_IMAGE_COUNT: We are yet to finish the quota of offline images before we can use images from the cache.
+        // * count >= STARTUP_IMAGE_COUNT but cachedImages.length == 0: Can happen when HTTP requests to r/EarthPorn fail. We fall back to display the backup images.
+        if (count < this.STARTUP_IMAGE_COUNT || cachedImages.length == 0) {
+            var order_index = count % this.BACKUP_IMAGES_COUNT;
+            this.displayImage(this.offlineImages[order[order_index]]);
         } else {
-            // TODO: Use variable cacheReadPos & cacheWritePos to prevent errors when fetch fails
-            var cachedImageNumber = cacheIndex - this.STARTUP_IMAGE_COUNT;
-            if (cachedImageNumber < 0) {
-                cachedImageNumber += cacheSize;
-            }
-            var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
-            this.displayImage(cachedImages[cachedImageNumber]);
+            this.displayImage(cachedImages[0]);
+            // Once we consume an image from the cache, remove it from the cache.
+            cachedImages.splice(0, 1);
+            localStorage.setItem("cachedImages", JSON.stringify(cachedImages));
         }
         var xmlHttp = new XMLHttpRequest();
         var parent = this;
         xmlHttp.onreadystatechange = function() {
+            // If the HTTP request is successful, then we store the image received into our cache. If the request is unsuccessful, the cache simply does not get updated.
             if (xmlHttp.readyState == 4 && xmlHttp.status == 200) {
                 var response = JSON.parse(xmlHttp.response);
                 // Fetch an image and cache it so it can be displayed quickly
                 var imageData = parent.parseImage(response, count);
-                var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
-                //var cacheIndex = Number(localStorage.getItem("cacheIndex"));
-                cachedImages[cacheIndex] = imageData;
+                cachedImages.push(imageData);
```

Comment:
Does this work even though cachedImages is not defined within this function? We had to use parent (this) earlier to solve this.

Reply:
cachedImages is in the same scope as parent is.  So it should be fine.

Reply:
Ah... good way of explaining it!